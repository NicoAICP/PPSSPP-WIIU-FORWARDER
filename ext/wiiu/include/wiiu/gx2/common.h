#pragma once

#define GX2_SCAN_BUFFER_ALIGNMENT      0x1000
#define GX2_SHADER_ALIGNMENT           0x100
#define GX2_CONTEXT_STATE_ALIGNMENT    0x100
#define GX2_UNIFORM_BLOCK_ALIGNMENT    0x100
#define GX2_DISPLAY_LIST_ALIGNMENT     0x20
#define GX2_VERTEX_BUFFER_ALIGNMENT    0x40
#define GX2_INDEX_BUFFER_ALIGNMENT     0x20

#define GX2_ENABLE                     TRUE
#define GX2_DISABLE                    FALSE

#define GX2_TRUE                       TRUE
#define GX2_FALSE                      FALSE

#ifndef GX2_COMP_SEL
#define GX2_COMP_SEL(c0, c1, c2, c3) (((c0) << 24) | ((c1) << 16) | ((c2) << 8) | (c3))
#ifdef __cplusplus
#include <functional>
#endif
#include <stdio.h>

#define _x 0
#define _y 1
#define _z 2
#define _w 3
#define _r 0
#define _g 1
#define _b 2
#define _a 3
#define _0 4
#define _1 5
#endif


#define GX2_DISABLE_WRAPS
