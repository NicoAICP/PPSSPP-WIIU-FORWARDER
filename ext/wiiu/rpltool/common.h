#ifndef COMMON_H
#define COMMON_H

#include <ctype.h>
#include <stdint.h>
#include <stdio.h>

typedef int8_t s8;
typedef int16_t s16;
typedef int32_t s32;

typedef uint8_t u8;
typedef uint16_t u16;
typedef uint32_t u32;

#define GET_ENUM_NAME(prefix, name, val) \
   case prefix##name: \
      return #name;

#define EMIT_ENUM_TO_STR_FUNC(name, cases) \
   static inline const char *name##_to_str(name type) { \
      switch (type) { cases } \
      return "unknown"; \
   }

#define GET_FLAG_NAME(prefix, name, val) \
   case prefix##name: \
      pos += snprintf(buffer + pos, sizeof(buffer) - pos, "%s%s", #name, seperator); \
      flags &= ~(1u << i); \
      break;

#define EMIT_FLAGS_TO_STR_FUNC(name, cases) \
   static inline const char *name##_to_str(int flags, const char *seperator) { \
      int i; \
      static char buffer[32 * 32]; \
      int pos = 0; \
      for (i = 0; i < 32; i++) \
         if ((1u << i) & flags) \
            switch (1u << i) { cases } \
      if (flags) \
         snprintf(buffer + pos, sizeof(buffer) - pos, "0x%08X", flags); \
      else if (pos) \
         buffer[pos - strlen(seperator)] = '\0'; \
      else \
         return "NULL"; \
      return buffer; \
   }

#define MAKE_MAGIC(a, b, c, d) (((a) << 24) | ((b) << 16) | ((c) << 8) | ((d) << 0))

#define log_var(X) printf("%-35s: 0x%0*X (%u)\n", #X, sizeof(X) * 2, (u32)(X), (u32)(X))
#define log_svar(X) printf("%-35s: 0x%0*X (%i)\n", #X, sizeof(X) * 2, (u32)(X), (s32)(X))
#define log_array(X) debug_array(X, sizeof(X), #X)
#define log_enum(type, X) printf("%-35s: 0x%X (%s)\n", #X, (X), type##_to_str(X))
#define log_flags(type, X) printf("%-35s: 0x%08X (%s)\n", #X, (X), type##_to_str(X, " | "))
#define log_magic(X) \
   printf("%-35s: '%c''%c''%c''%c' (0x%08X)\n", #X, (u32)(X) >> 24, (u32)(X) >> 16, (u32)(X) >> 8, (u32)(X), (u32)(X))
#define log_carray(X) debug_char_array(X, sizeof(X), #X)
#define log_str(X) printf("%-35s: \"%s\"\n", #X, (char *)(X))
#define log_off_str(X, data) printf("%-35s: 0x%08X \"%s\"\n", #X, (X), (char *)data + (X))
#define log_off_strs(X, data) \
   do { \
      printf("%-35s: 0x%08X ", #X, (X)); \
      const char *ptr = (char *)data + (X); \
      while (*ptr) { \
         ptr += printf("\"%s\" ", ptr) - 2; \
      } \
      printf("\n"); \
   } while (0)

static inline void dump_array(uint32_t *src, int size);
static inline void debug_array(uint32_t *src, int size, const char *name) {
   printf("%-35s:", name);
   dump_array(src, size);
}
static inline void debug_char_array(uint8_t *src, int size, const char *name) {
   int i;
   printf("%-35s: ", name);
   //   for (i = 0; i < size; i++)
   //      printf("%-2c ", isalnum(src[i])?src[i]: '?');
   //   printf("\n%-35s: ", "");
   for (i = 0; i < size; i++)
      printf("%02X ", src[i]);
   printf("\n");
}

static inline void dump_array(uint32_t *src, int size) {
   int i;
   size = (size / 4) - 1;
   printf("\n{");
   for (i = 0; i < size; i++) {
      if (!(i & 3))
         printf("\n   ");

      printf("0x%08X, ", src[i]);
   }

   if (!(i & 3))
      printf("\n   ");

   printf("0x%08X\n}\n", src[i]);
}

static inline u32 align_up(u32 x, u32 align) { return (x + (align - 1)) & ~(align - 1); }
static inline u32 align_down(u32 x, u32 align) { return x & ~(align - 1); }

#endif // COMMON_H
