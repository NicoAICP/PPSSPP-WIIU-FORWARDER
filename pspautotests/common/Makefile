TARGET_LIB = libcommon.a
EXTRA_TARGETS = libcommon_kernel.a
OBJS = common.o vram.o sysmem-imports.o
KERNEL_OBJS = common_kernel.o vram.o sysmem-imports.o

INCDIR = 
CFLAGS = -O0 -G 0 -g -Wall
CXXFLAGS = $(CFLAGS) -fno-exceptions -fno-rtti
ASFLAGS = $(CFLAGS)

LIBDIR =
LDFLAGS = -G0

PSPSDK=$(shell psp-config --pspsdk-path)
include $(PSPSDK)/lib/build.mak

%.o: %.S
	$(AS) $(ASFLAGS) -c -o $@ $<

common_kernel.o: common.c
	$(CC) $(CFLAGS) -DCOMMON_KERNEL=1 -c -o $@ $<

libcommon_kernel.a: $(KERNEL_OBJS)
	$(AR) cru $@ $(KERNEL_OBJS)
	$(RANLIB) $@
