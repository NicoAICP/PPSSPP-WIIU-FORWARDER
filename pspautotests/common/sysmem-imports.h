extern int sceKernelSetCompiledSdkVersion(int ver);
extern int sceKernelSetCompiledSdkVersion370(int ver);
extern int sceKernelSetCompiledSdkVersion380_390(int ver);
extern int sceKernelSetCompiledSdkVersion395(int ver);
extern int sceKernelSetCompiledSdkVersion401_402(int ver);
extern int sceKernelSetCompiledSdkVersion500_505(int ver);
extern int sceKernelSetCompiledSdkVersion507(int ver);
extern int sceKernelSetCompiledSdkVersion600_602(int ver);
extern int sceKernelSetCompiledSdkVersion603_605(int ver);
extern int sceKernelSetCompiledSdkVersion606(int ver);
extern int sceKernelGetCompiledSdkVersion();