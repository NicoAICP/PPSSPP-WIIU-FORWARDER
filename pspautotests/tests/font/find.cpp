#include <common.h>
#include <pspmodulemgr.h>
#include <pspiofilemgr.h>
#include <malloc.h>
extern "C" {
#include "libfont.h"
}

static FontLibraryHandle fontLib;
static FontLibraryHandle fontLibDone;
static int g_allocated = 0;

static void *allocQuiet(void *arg, u32 size) {
	g_allocated += size;
	void *p = malloc(size);
	//checkpoint("%08x = alloc(0x%x)", p, size);
	return p;
}

static void freeQuiet(void *arg, void *p) {
	//checkpoint("free(%08x)", p);
	free(p);
}

bool loadFontModule() {
	checkpointNext("Init");
	if (RUNNING_ON_EMULATOR) {
		return true;
	}
	SceUID fontModule = sceKernelLoadModule("libfont.prx", 0, NULL);
	if (fontModule <= 0) {
		printf("TEST ERROR: Unable to load libfont.prx\n");
		return false;
	}

	int status = -1;
	int result = sceKernelStartModule(fontModule, 0, NULL, &status, NULL);
	if (result != fontModule || status != 0) {
		printf("TEST ERROR: libfont.prx startup failed (%08x, %08x)\n", result, status);
		return false;
	}
	return true;
}

void testFindOptimum(const char *title, FontLibraryHandle lib, FontStyle *fontStyle, bool useError = true) {
	uint error = -1;
	g_allocated = 0;
	int index = sceFontFindFont(lib, fontStyle, useError ? &error : NULL);
	if (index < 0) {
		checkpoint("%s: Failed (%08x, %08x)", title, index, error);
	} else if (error != 0) {
		checkpoint("%s: Error (%08x, %08x)", title, index, error);
	} else {
		int allocated = g_allocated;
		checkpoint("%s: OK - %d (allocated %d)", title, index, allocated);
	}
}

void testBasicParams() {
	FontStyle style;
	memset(&style, 0, sizeof(style));

	checkpointNext("sceFontFindFont libs:");
	testFindOptimum("  Normal", fontLib, &style, true);
	testFindOptimum("  Invalid", 0, &style, true);
	testFindOptimum("  Done", fontLibDone, &style, true);

	checkpointNext("sceFontFindFont errors:");
	testFindOptimum("  Missing style", fontLib, NULL, true);
	// Crashes.
	//testFindOptimum("  No error pointer", fontLib, &style, false);
}

void testSizeMatch(const char *title, float h, float v, float hRes = 0.0f, float vRes = 0.0f) {
	FontStyle style;
	memset(&style, 0, sizeof(style));
	style.fontH = h;
	style.fontV = v;
	style.fontHRes = hRes;
	style.fontVRes = vRes;
	testFindOptimum(title, fontLib, &style);
}

void testSizeMatches() {
	checkpointNext("H match:");
	static const float sizes[] = {-1.0f, 0.01f, 1.0f, 3.0f, 5.0f, 6.5f, 7.0f, 8.0f, 8.5f, 8.5624f, 8.5625f, 9.0f, 10.0f, 10.125f, 11.0f, 12.0f, 13.0f, INFINITY};
	for (size_t i = 0; i < ARRAY_SIZE(sizes); ++i) {
		char temp[64];
		snprintf(temp, sizeof(temp), "  H=%f", sizes[i]);
		testSizeMatch(temp, sizes[i], 0.0f);
	}

	checkpointNext("V match:");
	for (size_t i = 0; i < ARRAY_SIZE(sizes); ++i) {
		char temp[64];
		snprintf(temp, sizeof(temp), "  V=%f", sizes[i]);
		testSizeMatch(temp, 0.0f, sizes[i]);
	}

	sceFontSetResolution(fontLib, 256.0f, 256.0f);
	testSizeMatch("  H=7.000000 (double res)", 7.0f, 0.0f);
	testSizeMatch("  H=3.500000 (double res)", 3.5f, 0.0f);
	testSizeMatch("  H=7.000000 (128.0f res)", 7.0f, 0.0f, 128.0f, 128.0f);
	sceFontSetResolution(fontLib, 128.0f, 128.0f);
}

void testBasicMatches() {
	FontStyle style;
	memset(&style, 0, sizeof(style));
	
	checkpointNext("General match:");
	style.fontFamily = 1;
	testFindOptimum("  Family = 1", fontLib, &style);
	style.fontFamily = 2;
	testFindOptimum("  Family = 2", fontLib, &style);
	style.fontFamily = 0;
	style.fontStyle = 1;
	testFindOptimum("  Style = 1", fontLib, &style);
	style.fontStyle = 5;
	testFindOptimum("  Style = 5", fontLib, &style);
	style.fontStyle = 6;
	testFindOptimum("  Style = 6", fontLib, &style);
	style.fontStyle = 103;
	testFindOptimum("  Style = 103", fontLib, &style);
	style.fontStyle = 0;
	style.fontLanguage = 1;
	testFindOptimum("  Language = 1", fontLib, &style);
	style.fontLanguage = 0;
	style.fontCountry = 1;
	testFindOptimum("  Country = 1", fontLib, &style);
	style.fontCountry = 0;
}

void testFilenameMatch(const char *title, const char *filename) {
	FontStyle style;
	memset(&style, 0, sizeof(style));
	strcpy(style.fontFileName, filename);
	testFindOptimum(title, fontLib, &style);
}

void testFilenameMatches() {
	checkpointNext("Filename match:");
	testFilenameMatch("  Exact", "ltn0.pgf");
	testFilenameMatch("  Partial", "ltn0.pg");
}

extern "C" int main(int argc, char *argv[]) {
	if (!loadFontModule()) {
		return 1;
	}

	FontNewLibParams libParams;
	memset(&libParams, 0, sizeof(libParams));
	libParams.allocFuncAddr = allocQuiet;
	libParams.freeFuncAddr = freeQuiet;
	libParams.numFonts = 4;

	uint error = -1;
	fontLib = sceFontNewLib(&libParams, &error);
	fontLibDone = sceFontNewLib(&libParams, &error);
	sceFontDoneLib(fontLibDone);

	testBasicParams();
	testSizeMatches();
	testBasicMatches();
	testFilenameMatches();

	sceFontDoneLib(fontLib);

	return 0;
}