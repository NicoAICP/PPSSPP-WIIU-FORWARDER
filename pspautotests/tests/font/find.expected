[x] Init

[x] sceFontFindFont libs:
[x]   Normal: OK - 0 (allocated 0)
[x]   Invalid: Error (00000000, 80460002)
[x]   Done: OK - 0 (allocated 0)

[x] sceFontFindFont errors:
[x]   Missing style: Error (00000000, 80460003)

[x] H match:
[x]   H=-1.000000: OK - 0 (allocated 0)
[x]   H=0.010000: Failed (ffffffff, 00000000)
[x]   H=1.000000: Failed (ffffffff, 00000000)
[x]   H=3.000000: Failed (ffffffff, 00000000)
[x]   H=5.000000: Failed (ffffffff, 00000000)
[x]   H=6.500000: Failed (ffffffff, 00000000)
[x]   H=7.000000: OK - 9 (allocated 0)
[x]   H=8.000000: Failed (ffffffff, 00000000)
[x]   H=8.500000: Failed (ffffffff, 00000000)
[x]   H=8.562400: Failed (ffffffff, 00000000)
[x]   H=8.562500: Failed (ffffffff, 00000000)
[x]   H=9.000000: Failed (ffffffff, 00000000)
[x]   H=10.000000: Failed (ffffffff, 00000000)
[x]   H=10.125000: OK - 0 (allocated 0)
[x]   H=11.000000: Failed (ffffffff, 00000000)
[x]   H=12.000000: Failed (ffffffff, 00000000)
[x]   H=13.000000: Failed (ffffffff, 00000000)
[x]   H=inf: Failed (ffffffff, 00000000)

[x] V match:
[x]   V=-1.000000: OK - 0 (allocated 0)
[x]   V=0.010000: Failed (ffffffff, 00000000)
[x]   V=1.000000: Failed (ffffffff, 00000000)
[x]   V=3.000000: Failed (ffffffff, 00000000)
[x]   V=5.000000: Failed (ffffffff, 00000000)
[x]   V=6.500000: Failed (ffffffff, 00000000)
[x]   V=7.000000: Failed (ffffffff, 00000000)
[x]   V=8.000000: Failed (ffffffff, 00000000)
[x]   V=8.500000: Failed (ffffffff, 00000000)
[x]   V=8.562400: Failed (ffffffff, 00000000)
[x]   V=8.562500: Failed (ffffffff, 00000000)
[x]   V=9.000000: Failed (ffffffff, 00000000)
[x]   V=10.000000: Failed (ffffffff, 00000000)
[x]   V=10.125000: Failed (ffffffff, 00000000)
[x]   V=11.000000: Failed (ffffffff, 00000000)
[x]   V=12.000000: Failed (ffffffff, 00000000)
[x]   V=13.000000: Failed (ffffffff, 00000000)
[x]   V=inf: Failed (ffffffff, 00000000)
[x]   H=7.000000 (double res): Failed (ffffffff, 00000000)
[x]   H=3.500000 (double res): OK - 9 (allocated 0)
[x]   H=7.000000 (128.0f res): OK - 9 (allocated 0)

[x] General match:
[x]   Family = 1: OK - 0 (allocated 0)
[x]   Family = 2: OK - 2 (allocated 0)
[x]   Style = 1: OK - 1 (allocated 0)
[x]   Style = 5: OK - 5 (allocated 0)
[x]   Style = 6: OK - 7 (allocated 0)
[x]   Style = 103: OK - 0 (allocated 0)
[x]   Language = 1: OK - 0 (allocated 0)
[x]   Country = 1: OK - 0 (allocated 0)

[x] Filename match:
[x]   Exact: OK - 1 (allocated 0)
[x]   Partial: Failed (ffffffff, 00000000)
